File: src/commands/schedule/pause-schedule.ts
/**
 * Pause/Resume Schedule Commands - SCHED-002
 *
 * Pauses or resumes a schedule in spec/schedules.json.
 */

import type { Command } from 'commander';
import { fileManager } from '../../utils/file-manager';
import { getSchedulesFilePath } from '../../utils/ensure-schedules-file';
import { output } from '../../utils/output';
import type {
  ScheduleNameOptions,
  ScheduleOperationResult,
  SchedulesData,
} from '../../types/schedule';

/**
 * Pauses a schedule by setting its status to 'paused'.
 *
 * @param options - Options containing the schedule name
 * @returns Result with success status
 */
export async function pauseSchedule(
  options: ScheduleNameOptions
): Promise<ScheduleOperationResult> {
  const cwd = options.cwd || process.cwd();
  const schedulesFile = getSchedulesFilePath(cwd);

  await fileManager.transaction<SchedulesData>(schedulesFile, async data => {
    const schedule = data.schedules[options.name];
    if (!schedule) {
      throw new Error(`Schedule '${options.name}' does not exist`);
    }

    if (schedule.status === 'paused') {
      throw new Error(`Schedule '${options.name}' is already paused`);
    }

    schedule.status = 'paused';
  });

  return { success: true };
}

/**
 * Resumes a schedule by setting its status to 'active'.
 *
 * @param options - Options containing the schedule name
 * @returns Result with success status
 */
export async function resumeSchedule(
  options: ScheduleNameOptions
): Promise<ScheduleOperationResult> {
  const cwd = options.cwd || process.cwd();
  const schedulesFile = getSchedulesFilePath(cwd);

  await fileManager.transaction<SchedulesData>(schedulesFile, async data => {
    const schedule = data.schedules[options.name];
    if (!schedule) {
      throw new Error(`Schedule '${options.name}' does not exist`);
    }

    if (schedule.status === 'active') {
      throw new Error(`Schedule '${options.name}' is already active`);
    }

    schedule.status = 'active';
  });

  return { success: true };
}

export function registerPauseScheduleCommand(program: Command): void {
  program
    .command('pause-schedule')
    .description('Pause a scheduled job')
    .argument('<name>', 'Schedule name to pause')
    .action(async (name: string) => {
      try {
        await pauseSchedule({ name });
        output.log(`✓ Schedule '${name}' paused successfully`);
      } catch (error: unknown) {
        output.error('✗ Failed to pause schedule:', (error as Error).message);
        process.exit(1);
      }
    });
}

export function registerResumeScheduleCommand(program: Command): void {
  program
    .command('resume-schedule')
    .description('Resume a paused scheduled job')
    .argument('<name>', 'Schedule name to resume')
    .action(async (name: string) => {
      try {
        await resumeSchedule({ name });
        output.log(`✓ Schedule '${name}' resumed successfully`);
      } catch (error: unknown) {
        output.error('✗ Failed to resume schedule:', (error as Error).message);
        process.exit(1);
      }
    });
}

File: rust/fspec-core/src/help/configs/pause_schedule.rs
//! `pause-schedule` help configuration — Rust port of the TS
//! `pause-schedule` Commander.js `--help` reference.
//!
//! Byte-for-byte parity with `node dist/index.js pause-schedule --help`
//! (captured at `rust/fspec/tests/fixtures/help/pause-schedule.txt`).

use super::super::{CommandArgument, CommandExample, CommandHelpConfig, CommonError};

const PREREQUISITES: &[&str] = &[
    "Schedule must exist in spec/schedules.json",
    "Schedule must currently have status \"active\"",
];

const ARGUMENTS: &[CommandArgument] = &[CommandArgument {
    name: "name",
    description: "The name of the schedule to pause (must be currently active)",
    required: true,
}];

const EXAMPLES: &[CommandExample] = &[
    CommandExample {
        command: "fspec pause-schedule nightly-review",
        description: Some("Pause the nightly-review schedule"),
        output: Some("✓ Schedule 'nightly-review' paused successfully"),
    },
    CommandExample {
        command: "fspec pause-schedule daily-tests",
        description: Some("Pause during a maintenance window"),
        output: Some("✓ Schedule 'daily-tests' paused successfully"),
    },
];

const COMMON_ERRORS: &[CommonError] = &[
    CommonError {
        error: "Schedule 'nightly-review' does not exist",
        fix: "Check the schedule name with 'fspec list-schedules'. Names are case-sensitive slugs.",
    },
    CommonError {
        error: "Schedule 'nightly-review' is already paused",
        fix: "The schedule is already paused. Use resume-schedule to reactivate it.",
    },
];

const RELATED: &[&str] = &[
    "resume-schedule - Resume a paused schedule",
    "list-schedules - List all schedules and their status",
    "remove-schedule - Permanently remove a schedule",
    "add-schedule - Create a new schedule",
];

const NOTES: &[&str] = &[
    "Pausing is non-destructive — the schedule configuration is preserved",
    "A paused schedule will not trigger until explicitly resumed",
    "Use list-schedules to verify the schedule status after pausing",
];

pub const CONFIG: CommandHelpConfig = CommandHelpConfig {
    name: "pause-schedule",
    description: "Pause a scheduled job by setting its status to \"paused\". The schedule remains in spec/schedules.json but will not trigger until resumed.",
    usage: Some("fspec pause-schedule <name>"),
    arguments: ARGUMENTS,
    options: &[],
    examples: EXAMPLES,
    related_commands: RELATED,
    when_to_use: Some("Use when you want to temporarily stop a schedule from triggering without removing it. Useful during maintenance windows, deployments, or when investigating issues."),
    when_not_to_use: Some("Do not use if you want to permanently remove the schedule. Use remove-schedule instead."),
    prerequisites: PREREQUISITES,
    common_patterns: &[],
    typical_workflow: Some("fspec list-schedules → fspec pause-schedule <name> (before maintenance) → perform maintenance → fspec resume-schedule <name> (after maintenance)"),
    common_errors: COMMON_ERRORS,
    notes: NOTES,
};
